### PR TITLE
Handle properly TCP connection closure

### PR DESCRIPTION
**Pull request by rveerama1**
In TCP active connection close, zephyr immediately moves to CLOSED state from TIME_WAIT. Which is causing last ACK message to be dropped. Peer device keep on sending FIN+ACK messages.

TIMEWAIT should be 2 * MSL (Maximum Segment Lifetime). Now I am introduced only 100ms. It needs to be properly calculated. Connection waits for timewait timer. In mean time ACK will be properly sent and connection will be terminated properly. Issue noticed in https://github.com/zephyrproject-rtos/zephyr/issues/4324.

### CODE REVIEW

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -1533,6 +1559,9 @@ int net_context_connect(struct net_context *context,
 	if (ret < 0) {
 		return ret;
 	}
+	/* TODO: Only in-case of ret = -ENOENT, should we wait for free
+	 * available connection handler until given 'timeout' ?
+	 */
```
Comment (pfl) [inline subsys/net/ip/net_context.c]: This comment is for another (API) matter, let's have it in a different patch addressing that issue.

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -49,6 +49,9 @@
 #define ACK_TIMEOUT K_SECONDS(1)
 #endif
 
+/* TODO: It should be 2 * MSL (Maximum segment lifetime) */
+#define TIMEWAIT_TIMEOUT MSEC(100)
+
```
Comment (pfl) [inline subsys/net/ip/net_context.c]: This is another issue discovered, indeed. Will 100 ms be enough for the current cases or should the timeout be longer? 100 ms sounds awfully short for some reason.

**Review by pfl — CHANGES_REQUESTED** [review]
The commit message could be a bit clearer by first stating that the states follow each other in sequence, with the last state erasing a properly created ACK as the context is freed. Likewise, go through the solution in a similar manner, so that one learns how the added timer adds value.

File: samples/net/zperf/prj_netusb.conf [hunk 3]
```diff
@@ -6,8 +6,8 @@ CONFIG_NET_UDP=y
 CONFIG_NET_TCP=y
 CONFIG_NET_STATISTICS=y
 
-CONFIG_NET_PKT_RX_COUNT=100
-CONFIG_NET_PKT_TX_COUNT=100
```
Comment (rveerama1) [inline samples/net/zperf/prj_netusb.conf]: These changes are not related to TCP issue, but sanitycheck fails to build because of extra 272Bytes in RAM.

**Comment by rveerama1** [comment]
Fixed @pfl comments.

**Review by pfl — APPROVED** [review]

File: subsys/net/ip/net_context.c [hunk 4]
```diff
@@ -301,6 +304,26 @@ static void handle_ack_timeout(struct k_work *work)
...
+	struct net_tcp *tcp = CONTAINER_OF(work, struct net_tcp,
+					   timewait_timer);
+
+	NET_DBG("Timewait expired in %dms", TIMEWAIT_TIMEOUT);
+
+	if (net_tcp_get_state(tcp) == NET_TCP_TIME_WAIT) {
+		net_tcp_change_state(tcp, NET_TCP_CLOSED);
+
+		if (tcp->context->recv_cb) {
+			tcp->context->recv_cb(tcp->context, NULL, 0,
```
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: @rveerama1 : Any reason why you put this code here? All the states like FIN_WAIT, TIME_WAIT are internal TCP states. From the application point of view, the connection is closed when first (valid) FIN is received. The rest is again just the internal TCP connection shutdown sequence. So, calling recv_cb(NULL) at the end of TIME_WAIT expiration seems like too late. I'm considering submitting a patch to remove it from here, unless you have better suggestions. Thanks.

Comment (rveerama1) [reply subsys/net/ip/net_context.c]: I wasn't sure what kind of corner cases hits this original code without the timer. If you have any cases which is causing issues with it, please send a fix.

**Pull request merged**